shared/lib/accounts/accounts.ts

```diff
@@ -35,46 +50,192 @@ export function getUniqueAccountName(
+  never
+>;
+
+export class MultichainWalletSnapClient implements WalletSnapClient {
+  #snapId: SUPPORTED_WALLET_SNAP_ID;
+
+  #snapKeyring: SnapKeyring;
+
+  #scopes: CaipChainId[];
+
+  #client: KeyringInternalSnapClient;
 
-    await Promise.allSettled(
-      discovered.map(async (discoveredAccount) => {
-        const options = {
-          derivationPath: discoveredAccount.derivationPath,
-          entropySource,
-        };
-
-        // TODO: Use `withKeyring` instead of using the keyring directly.
-        await snapKeyring.createAccount(snapId, options, {
-          displayConfirmation: false,
-          displayAccountNameSuggestion: false,
-          setSelectedAccount: false,
-        });
+  #messenger: MultichainWalletSnapClientMessenger;
+
+  constructor(
+    snapId: SUPPORTED_WALLET_SNAP_ID,
+    scopes: CaipChainId[],
+    snapKeyring: SnapKeyring,
+    messenger: MultichainWalletSnapClientMessenger,
+  ) {
+    this.#snapId = snapId;
+    this.#snapKeyring = snapKeyring;
+
+    this.#scopes = scopes;
+
+    this.#messenger = messenger;
+
+    this.#client = new KeyringInternalSnapClient({
+      snapId,
+      // @ts-expect-error TODO: Resolve mismatch between base-controller versions.
+      messenger: messenger.getRestricted({
+        name: 'KeyringInternalSnapClient',
+        allowedActions: ['SnapController:handleRequest'],
+        allowedEvents: [],
       }),
+    });
+  }
+
+  getSnapId(): SnapId {
+    return this.#snapId;
+  }
+
+  async createAccount(
+    { accountNameSuggestion, ...options }: WalletSnapOptions,
+    internalOptions?: SnapKeyringInternalOptions,
+  ): Promise<KeyringAccount> {
+    // Automatically name the account if not provided.
+    const autoAccountNameSuggestion = accountNameSuggestion
+      ? { accountNameSuggestion }
+      : { accountNameSuggestion: await this.getNextAvailableAccountName() };
+
+    // TODO: Use `withKeyring` instead of using the keyring directly.
+    return await this.#snapKeyring.createAccount(
+      this.#snapId,
+      {
+        ...options,
+        // TODO: Stop forwarding the account name to the Snap, we should make this
+        // an internal option at some point.
+        ...autoAccountNameSuggestion,
+      },
+      internalOptions,
+    );
+  }
+
+  async getNextAvailableAccountName(
+    options: SnapAccountNameOptions = {},
+  ): Promise<string> {
+    return getNextAvailableSnapAccountName(
+      async () => {
+        return this.#messenger.call(
+          'AccountsController:getNextAvailableAccountName',
+          KeyringTypes.snap,
+        );
+      },
+      this.#snapId,
+      options,
+    );
+  }
+
+  async #discoverAccountsOn(
+    entropySource: string,
+    groupIndex: number,
+  ): Promise<DiscoveredAccount[]> {
+    return await this.#client.discoverAccounts(
+      this.#scopes,
+      entropySource,
+      groupIndex,
     );
   }
+
+  async discoverAccounts(entropySource: string) {
+    for (let index = 0; ; index++) {
```

Comment:
Instead of sequentially resolving each promise, do you think is possible to batch them and use `Promise.all`? I say batch because we should probably not use it with 100 accounts but instead do it in groups of 10 or 20, we can use a mutex to make sure the execution of `getNextAvailableAccountName` is safe

Reply:
I'd go sequentially for now to keep things simple.

But I can add a `TODO` for this improvement maybe (your suggestion make sense, so I'd like to have this somewhere still).